File: ios/DSHAnywhere/Core/Store/DSHProfileStore.swift
import Foundation

/// The Macs this iPhone is paired with.
///
/// The app kept exactly one profile, so pairing a second Mac silently replaced
/// the first and a user with two Macs could never reach both. This stores an
/// ordered list plus the active machine, and folds the old single-profile key
/// into the list on first read so an existing install keeps its pairing.
///
/// Device tokens already live in the Keychain keyed by `deviceId`, so several
/// machines' credentials can coexist without further changes.
public struct DSHProfileStore: @unchecked Sendable {
    public static let profilesKey = "dsh-anywhere.relay-profiles"
    public static let activeMachineKey = "dsh-anywhere.active-machine-id"
    /// Written by the single-profile version; read once, for migration.
    public static let legacyProfileKey = "dsh-anywhere.relay-profile"

    // `UserDefaults` is documented as thread-safe; the unchecked conformance is
    // only needed because the type itself is not marked Sendable.
    private let defaults: UserDefaults

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    public var profiles: [DSHRemoteProfile] { load() }

    public var activeMachineId: String? {
        defaults.string(forKey: Self.activeMachineKey) ?? load().first?.machineId
    }

    public var activeProfile: DSHRemoteProfile? {
        let all = load()
        guard let id = activeMachineId else { return all.first }
        return all.first { $0.machineId == id } ?? all.first
    }

    /// Adds or replaces one machine. Re-pairing the same Mac updates it in place
    /// instead of appending a duplicate.
    @discardableResult
    public func upsert(_ profile: DSHRemoteProfile, makeActive: Bool = true) -> [DSHRemoteProfile] {
        var all = load().filter { $0.machineId != profile.machineId }
        all.append(profile)
        save(all)
        if makeActive { setActive(profile.machineId) }
        return all
    }

    /// Removes one machine, moving the active selection to a remaining Mac so a
    /// removal can never leave the app pointing at nothing.
    @discardableResult
    public func remove(_ machineId: String) -> [DSHRemoteProfile] {
        let all = load().filter { $0.machineId != machineId }
        save(all)
        if activeMachineId == machineId {
            if let next = all.first { setActive(next.machineId) }
            else { defaults.removeObject(forKey: Self.activeMachineKey) }
        }
        return all
    }

    public func setActive(_ machineId: String) {
        defaults.set(machineId, forKey: Self.activeMachineKey)
    }

    private func load() -> [DSHRemoteProfile] {
        if let data = defaults.data(forKey: Self.profilesKey),
           let decoded = try? JSONDecoder().decode([DSHRemoteProfile].self, from: data) {
            return decoded
        }
        guard let legacy = defaults.data(forKey: Self.legacyProfileKey),
              let profile = try? JSONDecoder().decode(DSHRemoteProfile.self, from: legacy) else {
            return []
        }
        save([profile])
        setActive(profile.machineId)
        return [profile]
    }

    private func save(_ profiles: [DSHRemoteProfile]) {
        guard let data = try? JSONEncoder().encode(profiles) else { return }
        defaults.set(data, forKey: Self.profilesKey)
    }
}

File: android/app/src/main/kotlin/com/dshanywhere/core/store/DSHProfileStore.kt
package com.dshanywhere.core.store

import android.content.Context
import com.dshanywhere.core.network.DSHRemoteProfile
import com.dshanywhere.core.protocol.DSHJson

/**
 * The Macs this device is paired with. Mirrors
 * ios/DSHAnywhere/Core/Store/DSHProfileStore.swift: an ordered list plus the
 * active machine; device tokens live in the token store keyed by deviceId, so
 * several machines' credentials can coexist without further changes.
 */
class DSHProfileStore(context: Context) {
    private val prefs = context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)

    val profiles: List<DSHRemoteProfile> get() = load()

    val activeMachineId: String?
        get() = prefs.getString(KEY_ACTIVE, null) ?: load().firstOrNull()?.machineId

    val activeProfile: DSHRemoteProfile?
        get() {
            val all = load()
            val id = activeMachineId ?: return all.firstOrNull()
            return all.firstOrNull { it.machineId == id } ?: all.firstOrNull()
        }

    /** Adds or replaces one machine, keeping pairing order stable. */
    fun upsert(profile: DSHRemoteProfile, makeActive: Boolean = true): List<DSHRemoteProfile> {
        val all = load().filterNot { it.machineId == profile.machineId } + profile
        save(all)
        if (makeActive) setActive(profile.machineId)
        return all
    }

    /** Removes one machine, moving the active selection to a remaining Mac. */
    fun remove(machineId: String): List<DSHRemoteProfile> {
        val all = load().filterNot { it.machineId == machineId }
        save(all)
        if (activeMachineId == machineId) {
            val next = all.firstOrNull()
            if (next != null) setActive(next.machineId)
            else prefs.edit().remove(KEY_ACTIVE).apply()
        }
        return all
    }

    fun setActive(machineId: String) {
        prefs.edit().putString(KEY_ACTIVE, machineId).apply()
    }

    private fun load(): List<DSHRemoteProfile> {
        val raw = prefs.getString(KEY_PROFILES, null) ?: return emptyList()
        return runCatching {
            DSHJson.decodeFromString(
                kotlinx.serialization.builtins.ListSerializer(DSHRemoteProfile.serializer()), raw,
            )
        }.getOrDefault(emptyList())
    }

    private fun save(profiles: List<DSHRemoteProfile>) {
        val raw = DSHJson.encodeToString(
            kotlinx.serialization.builtins.ListSerializer(DSHRemoteProfile.serializer()), profiles,
        )
        prefs.edit().putString(KEY_PROFILES, raw).apply()
    }

    companion object {
        const val PREFERENCES = "dsh-anywhere"
        const val KEY_PROFILES = "dsh-anywhere.relay-profiles"
        const val KEY_ACTIVE = "dsh-anywhere.active-machine-id"
    }
}
